### PR TITLE
logql: handle negative offset

### LINKED ISSUE #7476: Offset modifier doesn't accept negative values
**Describe the bug**
Negative offset modifiers in logql queries from Grafana throw the following exception:
`parse error at line 2, col 408: syntax error: unexpected -, expecting DURATION`

**To Reproduce**
Logql query with negative offset example (from prometheus documentation: https://prometheus.io/docs/prometheus/latest/querying/basics/#offset-modifier):
`rate(http_requests_total[5m] offset -1w)`

My current query:
`sum by (source) (count_over_time({logset="api_dataset"}[1d] offset -1d))`

**Expected behavior**
Same as promql negative offsets

**Environment:**
 - Infrastructure: docker-swarm

### PR DESCRIPTION
**Pull request by cunyat**
**What this PR does / why we need it**:
Handles negative offset in logql correctly

**Which issue(s) this PR fixes**:
Fixes #7476

**Special notes for your reviewer**:
I'm not really sure that I've implemented the best solution as it maybe tries to parse duration in many other cases. 

**Checklist**
- [x] Reviewed the `CONTRIBUTING.md` guide
- [ ] Documentation added
- [x] Tests updated
- [ ] `CHANGELOG.md` updated
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/upgrading/_index.md`

### CODE REVIEW

File: pkg/logql/syntax/lex.go [hunk 1]
```diff
@@ -151,6 +151,12 @@ func (l *lexer) Lex(lval *exprSymType) int {
 
 		lval.str = numberText
 		return NUMBER
+	case '-': // handle negative durations
+		tokenText := l.TokenText()
+		if duration, ok := tryScanDuration(tokenText, &l.Scanner); ok {
+			lval.duration = duration
```
[the remark this answers is not in the document]
Comment (cunyat) [reply pkg/logql/syntax/lex.go:157]: The point is to parse correctly negative offset. When it founds `-` it looks ahead for a duration.
Also, durations between `[]` brackets are parsed in other case of that function [see line 177](https://github.com/grafana/loki/blob/881b5193f6ecd5623c7f9cc48d00d52e68ae4ebd/pkg/logql/syntax/lex.go#L177). 

Do you think there is a better approach?

[the remark this answers is not in the document]
Comment (jeschkies) [reply pkg/logql/syntax/lex.go:157]: Ah, I had to dig into the implementation. so `TokenText` will be `-123` or whatever the duration is. Cool. Could you rebase on `main` so we can merge it?

[the remark this answers is not in the document]
Comment (cunyat) [reply pkg/logql/syntax/lex.go:163]: Ready :)

[the remark this answers is not in the document]
Comment (cunyat) [reply pkg/logql/syntax/lex.go:163]: @jeschkies any updates on this? I'm trying to keep it updated with `main` :)

**Comment by JStickler** [comment]
Removed myself as reviewer as I am not a developer but a technical writer and this PR does not touch any of the documentation files.

**Review by jeschkies — APPROVED** [review]

File: pkg/logql/syntax/lex.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (jeschkies) [reply pkg/logql/syntax/lex.go:163]: Sorry for the very long delay. I lost track of it.

**Pull request merged**